File: docs/breez-sdk/snippets/go/reserve_leaves.go
package example

import (
	"log"
	"math/big"

	"github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
)

func PrepareSendPaymentReserveLeaves(sdk *breez_sdk_spark.BreezSdk) (*breez_sdk_spark.PrepareSendPaymentResponse, error) {
	// ANCHOR: prepare-send-payment-reserve-leaves
	paymentRequest := "<payment request>"
	amountSats := new(big.Int).SetInt64(50_000)
	reserveLeaves := true

	request := breez_sdk_spark.PrepareSendPaymentRequest{
		PaymentRequest:    paymentRequest,
		Amount:            &amountSats,
		TokenIdentifier:   nil,
		ConversionOptions: nil,
		FeePolicy:         nil,
		ReserveLeaves:     &reserveLeaves,
	}
	response, err := sdk.PrepareSendPayment(request)
	if err != nil {
		return nil, err
	}

	// The reservation ID can be used to cancel the reservation if needed
	if response.ReservationId != nil {
		log.Printf("Reservation ID: %v", *response.ReservationId)
	}

	// Send payment as usual using the prepare response
	// sdk.SendPayment(breez_sdk_spark.SendPaymentRequest{PrepareResponse: response, Options: nil, IdempotencyKey: nil})
	// ANCHOR_END: prepare-send-payment-reserve-leaves
	return &response, nil
}

func CancelPrepareSendPayment(sdk *breez_sdk_spark.BreezSdk) error {
	// ANCHOR: cancel-prepare-send-payment
	reservationId := "<reservation id from prepare response>"

	request := breez_sdk_spark.CancelPrepareSendPaymentRequest{
		ReservationId: reservationId,
	}
	err := sdk.CancelPrepareSendPayment(request)
	if err != nil {
		return err
	}
	// ANCHOR_END: cancel-prepare-send-payment
	return nil
}

File: docs/breez-sdk/snippets/python/src/reserve_leaves.py
import logging
from breez_sdk_spark import (
    BreezSdk,
    CancelPrepareSendPaymentRequest,
    PrepareSendPaymentRequest,
)


async def prepare_send_payment_reserve_leaves(sdk: BreezSdk):
    # ANCHOR: prepare-send-payment-reserve-leaves
    payment_request = "<payment request>"
    amount_sats = 50_000
    try:
        prepare_response = await sdk.prepare_send_payment(
            request=PrepareSendPaymentRequest(
                payment_request=payment_request,
                amount=amount_sats,
                token_identifier=None,
                conversion_options=None,
                fee_policy=None,
                reserve_leaves=True,
            )
        )

        # The reservation ID can be used to cancel the reservation if needed
        if prepare_response.reservation_id is not None:
            logging.debug(f"Reservation ID: {prepare_response.reservation_id}")

        # Send payment as usual using the prepare response
        # await sdk.send_payment(
        #     request=SendPaymentRequest(prepare_response=prepare_response)
        # )
    except Exception as error:
        logging.error(error)
        raise
    # ANCHOR_END: prepare-send-payment-reserve-leaves


async def cancel_prepare_send_payment(sdk: BreezSdk):
    # ANCHOR: cancel-prepare-send-payment
    reservation_id = "<reservation id from prepare response>"
    try:
        await sdk.cancel_prepare_send_payment(
            request=CancelPrepareSendPaymentRequest(
                reservation_id=reservation_id,
            )
        )
    except Exception as error:
        logging.error(error)
        raise
    # ANCHOR_END: cancel-prepare-send-payment
